samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -319,6 +319,19 @@ private Map<TaskName, Checkpoint> restoreStores() throws InterruptedException {
       Map<TaskName, Checkpoint> newTaskCheckpoints = initRestoreAndNewCheckpointFuture.get();
       taskCheckpoints.putAll(newTaskCheckpoints);
       LOG.debug("Post init and restore checkpoints is: {}. NewTaskCheckpoints are: {}", taskCheckpoints, newTaskCheckpoints);
+      // Stop all persistent stores after restoring. Certain persistent stores opened in BulkLoad mode are compacted
+      // on stop, so paralleling stop() also parallelizes their compaction (a time-intensive operation).
```

Comment:
This is no longer parallel. That can be a big performance bottleneck for jobs restoring from Kafka changelogs since compaction can be slow.

Reply:
You're right. Fixed it by calling taskRestoreManager.close() after each taskRestoreManager.restore() completes (with or without expcetion). Also added test to verify that the close is not prematurely called.